In [30]:
import pstats

p =pstats.Stats('results/cprofile4.txt')
p.strip_dirs().sort_stats(-1).print_stats()

p.sort_stats('cumtime').print_stats(100)

Mon Dec 14 21:59:27 2015    results/cprofile4.txt

         825703 function calls (724515 primitive calls) in 145.303 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <decorator-gen-0>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-100>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-101>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-101>:1(_not_implemented_for)
        1    0.000    0.000    0.000    0.000 <decorator-gen-102>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-103>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-103>:1(_not_implemented_for)
        1    0.000    0.000    0.000    0.000 <decorator-gen-104>:1(<module>)
        1    0.000    0.000    0.000    0.000 <decorator-gen-105>:1(<module>)
        1    0.000    0.000    0.000    0.000 <

In [33]:
import line_profiler

In [52]:
lstats = line_profiler.load_stats("run_simulation.py.lprof")
line_profiler.show_text(lstats.timings, lstats.unit)

Timer unit: 1e-06 s

Total time: 28.8485 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/graph_local_classes.py
Function: sample_edge_first_event_only at line 241

Line #      Hits         Time  Per Hit   % Time  Line Contents
   241                                               @profile
   242                                               def sample_edge_first_event_only(self, edge, time):
   243   3366778      3823208      1.1     13.3          index = self.structure.edges.index(edge)
   244                                                   
   245                                                   # does it occur?
   246   3366778      3972137      1.2     13.8          occurs = np.random.rand() < self.params.p
   247   3366778      2130701      0.6      7.4          if not occurs:
   248    672609       394393      0.6      1.4              return None
   249                                                   
   250                          

Timer unit: 1e-06 s

Total time: 2.90067 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/graph_local_classes.py
Function: sample_edge_first_event_only at line 241

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   241                                               @profile
   242                                               def sample_edge_first_event_only(self, edge, time):
   243    326725       381856      1.2     13.2          index = self.structure.edges.index(edge)
   244                                                   
   245                                                   # does it occur?
   246    326725       398866      1.2     13.8          occurs = np.random.rand() < self.params.p
   247    326725       208333      0.6      7.2          if not occurs:
   248     65391        39652      0.6      1.4              return None
   249                                                   
   250                                                   # how many events?
   251    261334       623035      2.4     21.5          num_events = np.random.poisson(lam=self.params.mu[index])
   252    261334       187189      0.7      6.5          if num_events == 0:
   253    130382        79183      0.6      2.7              return None
   254                                                   
   255                                                   # when do those events occur?
   256    130952       870888      6.7     30.0          event_time = time + np.random.exponential(scale=self.params.r[index]/num_events, size=1)
   257                                                   # event_times.sort()
   258                                                   # return [(t, edge[1]) for t in event_times]
   259                                                   # import ipdb; ipdb.set_trace()
   260    130952       111672      0.9      3.8          return (event_time, edge[1])

Total time: 12.1135 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/graph_local_classes.py
Function: _sample_solely_first_events at line 317

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   317                                               @profile
   318                                               def _sample_solely_first_events(self, max_time=4.0):
   319    121600       121897      1.0      1.0          pending = []
   320    121600       187755      1.5      1.5          heapq.heappush(pending,(self.init_time, self.init_node))
   321                                                   # pending = [(self.init_time, self.init_node)]
   322    121600       140176      1.2      1.2          self._all_events = []
   323    121600       386441      3.2      3.2          self._first_events = {node: np.inf for node in self.structure.nodes}
   324    121600       111870      0.9      0.9          short_circuit = False
   325    121600       144643      1.2      1.2          processed_nodes = set()
   326    121600       178827      1.5      1.5          structure_nodes = set(self.structure.nodes)
   327    342096       395455      1.2      3.3          while len(pending) > 0:
   328                                                       # time, node = pending.pop(0)
   329                                                       # import ipdb; ipdb.set_trace()
   330    249445       329548      1.3      2.7              time, node = heapq.heappop(pending)
   331    249445       452177      1.8      3.7              if time >= max_time:
   332     14038        13882      1.0      0.1                  break
   333                                           
   334                                                       #self._all_events.append((time, node))
   335    235407       392782      1.7      3.2              if time < self._first_events[node]:
   336    234195       254233      1.1      2.1                  self._first_events[node] = time
   337                                           
   338    235407       238530      1.0      2.0              if node in processed_nodes:
   339      1212         1117      0.9      0.0                  continue
   340                                                       else:
   341    234195       252685      1.1      2.1                  processed_nodes.add(node)
   342    234195       237891      1.0      2.0              if processed_nodes==structure_nodes:
   343                                                           # this only works if it is first event only
   344     14911        15004      1.0      0.1                  break
   345                                           
   346    219284       251248      1.1      2.1              children_edges = self.structure.child_edges[node]
   347                                           # this goes to each of the children_edges of the node and initiates events along that edge
   348                                                       # import ipdb; ipdb.set_trace()
   349    562930       602900      1.1      5.0              for edge in children_edges:
   350    343646       380228      1.1      3.1                  if edge[1] in processed_nodes:
   351     16921        15855      0.9      0.1                      continue
   352                                                           else:
   353    326725      5581106     17.1     46.1                      child_events = self.sample_edge_first_event_only(edge, time)
   354                                                               # import ipdb; ipdb.set_trace()
   355    326725       341669      1.0      2.8                      if not child_events:
   356    195773       188015      1.0      1.6                          continue
   357                                                               # import ipdb; ipdb.set_trace()
   358    130952       236442      1.8      2.0                      heapq.heappush(pending,child_events)
   359                                                           # import ipdb; ipdb.set_trace()
   360                                                           # pending.sort()
   361                                                       
   362                                                   #self._all_events.sort()
   363                                                   #self._compute_first_events()
   364                                                   #return np.array(self._first_events)
   365    121600       661143      5.4      5.5          return np.array([self._first_events[node] for node in self.structure.nodes])

Total time: 21.2028 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: p_graph_given_d at line 39

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    39                                               @profile
    40                                               def p_graph_given_d(self,graphs,options):
    41                                                   # sets a catch for all numerical warnings to be treated as errors
    42                                                   # np.seterr(all='raise')
    43                                                   # np.seterr(under='raise')
    44         4           69     17.2      0.0          np.seterr(over='raise')
    45                                                   """
    46                                                   options includes 
    47                                                   sparsity: the sparsity argument for the prior 
    48                                                   param_sample_size: the number of parameters to be sampled
    49                                                   stigma_sample_size: the number of internal states to be sampled
    50                                                   data_sets: the different data_sets to be evaluated as likelihoods
    51                                                   data_p: the probability of each data_set in data_sets
    52                                                   num_data_samps: number of samples of observed data to be considered in the liklihood
    53                                                   """
    54         4           10      2.5      0.0          self.graphs = graphs
    55                                           
    56                                                   # Note: some day this will need to change if graphs is made to be an iterator and not a list
    57         4            7      1.8      0.0          self.max_graph = self.graphs[0]
    58         4            8      2.0      0.0          self.options = options
    59         4            6      1.5      0.0          num_graphs = len(graphs)
    60                                           
    61                                                   # loglikelihood = np.empty(len(self.graphs))
    62         4            8      2.0      0.0          num_params= options["param_sample_size"]
    63                                           
    64                                                   # generate 1 complete graph with many data structures shared beneath it
    65         4           11      2.8      0.0          self.gs_out = GraphStructure.from_networkx(sub_graph_from_edge_type(self.max_graph,
    66         4         4776   1194.0      0.0              edge_types=["observed"]))
    67                                                   
    68         4            8      2.0      0.0          self.gs_in = [GraphStructure.from_networkx(sub_graph_from_edge_type(graph,
    69         4       820287 205071.8      3.9              edge_types=["hidden_sample"])) for graph in graphs]
    70                                           
    71         4          390     97.5      0.0          max_graph_params = GraphParams.from_networkx(self.max_graph)
    72                                                   
    73         4          668    167.0      0.0          self.param_list = [max_graph_params.sample() for x in range(num_params)]
    74                                           
    75         4            9      2.2      0.0          if self.options["parallel"]:            
    76                                                       loglikelihood_by_param = np.array(Parallel(n_jobs = -1, 
    77                                                           backend = "multiprocessing", verbose = 20)(
    78                                                           delayed(self.subgraph_cross_entropy)(
    79                                                               max_graph_params.from_dict(params)) for params in self.param_list))
    80                                                   else:
    81         4            9      2.2      0.0              loglikelihood_by_param = np.array([self.subgraph_cross_entropy(max_graph_params.from_dict(params)) 
    82         4     20291329 5072832.2     95.7                  for params in self.param_list])
    83                                                   
    84         4          337     84.2      0.0          loglikelihood = logmeanexp(loglikelihood_by_param,axis=0)
    85                                                   
    86         4            8      2.0      0.0          sparsity = options["sparsity"]
    87         4        84847  21211.8      0.4          logposterior = self.logposterior_from_loglik_logsparseprior(loglikelihood,sparsity=sparsity)
    88                                           
    89         4           31      7.8      0.0          return graphs,np.exp(logposterior),loglikelihood,self.options,self.param_list

Total time: 20.2912 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: subgraph_cross_entropy at line 91

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    91                                               @profile
    92                                               def subgraph_cross_entropy(self,max_graph_params):
    93         4            6      1.5      0.0          n = self.options["num_data_samps"]
    94         4           17      4.2      0.0          q = np.array(self.options["data_probs"])
    95         4           35      8.8      0.0          δ = np.array(self.options["data_sets"])
    96         4         2470    617.5      0.0          gp_out = max_graph_params.subgraph_copy(self.gs_out.edges)
    97                                           
    98                                           
    99                                                   # note that q*approx_loglik_from_hidden_states should be vector)
   100         4     20288646 5072161.5    100.0          return np.array([n*np.dot(q,self.approx_loglik_from_hidden_states(δ,graph,max_graph_params,gp_out,g_idx)) for g_idx,graph in enumerate(self.graphs)])

Total time: 0 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: gen_iter_simulations_first_only at line 102

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   102                                               @profile
   103                                               def gen_iter_simulations_first_only(self,gs_in,gp_in,K):
   104                                                   # builds a simulation object and then samples returning an M lengthed generator
   105                                                   inner_simul = InnerGraphSimulation(gs_in, gp_in)
   106                                                   return inner_simul.sample_iter_solely_first_events(K)

Total time: 18.7658 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: gen_simulations_first_only at line 108

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   108                                               @profile
   109                                               def gen_simulations_first_only(self,gs_in,gp_in,K):
   110                                                   # builds a simulation object and then samples returning an M lengthed generator
   111      1216         6459      5.3      0.0          inner_simul = InnerGraphSimulation(gs_in, gp_in)
   112      1216     18759302  15427.1    100.0          return inner_simul.sample_solely_first_events(K)

Total time: 20.2659 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: approx_loglik_from_hidden_states at line 114

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   114                                               @profile
   115                                               def approx_loglik_from_hidden_states(self,data_sets,graph,max_graph_params,gp_out,g_idx):
   116      1216         1782      1.5      0.0          K = self.options["stigma_sample_size"]
   117                                           
   118                                                   # gs_in = GraphStructure.from_networkx(sub_graph_from_edge_type(graph,
   119                                                   #     edge_types=["hidden_sample"]))
   120      1216       762425    627.0      3.8          gp_in = max_graph_params.subgraph_copy(self.gs_in[g_idx].edges)
   121                                           
   122                                                   # hidden_states_iter = self.gen_iter_simulations_first_only(self.gs_in[g_idx],gp_in,K)
   123      1216     18776918  15441.5     92.7          hidden_states_iter = self.gen_simulations_first_only(self.gs_in[g_idx],gp_in,K)
   124                                           
   125      1216         6244      5.1      0.0          temp_array = np.zeros(shape=(K,data_sets.shape[0]))
   126                                                   # import ipdb; ipdb.set_trace()
   127                                                   # for idx, hidden_state_sample in enumerate(hidden_states_iter):
   128                                                   #     temp_array[idx,:] = np.array([self.loglik_with_hidden_states(data_set,hidden_state_sample,gp_out) for data_set in data_sets])
   129                                                   # import ipdb; ipdb.set_trace()
   130                                           
   131                                                   # return value shape should give hidden_sample (first idx) by data_set (idx) summed loglik
   132      1216       595016    489.3      2.9          temp_array = self.loglik_with_hidden_states_vectorized(data_sets,hidden_states_iter,gp_out)
   133                                                   # import IPython; IPython.embed()
   134                                                   # import ipdb; ipdb.set_trace()
   135      1216       123515    101.6      0.6          return logmeanexp(temp_array,axis=0)

Total time: 0.584242 s
File: /Users/cocosci/Dropbox/Work/Berkeley/Projects/Time/hidden_structure_inference/lib/likelihood_calculations_shared_params.py
Function: loglik_with_hidden_states_vectorized at line 137

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   137                                               @profile
   138                                               def loglik_with_hidden_states_vectorized(self,data_sets,hidden_states,gp_out):
   139                                           
   140                                                   # K = hidden_states.shape[0]
   141                                                   # temp_array = np.zeros(shape=(K,data_sets.shape[0]))
   142                                           
   143                                                   # creates new axis so that it broadcasts correctly
   144                                           
   145      1216         3349      2.8      0.6          h_states = hidden_states[:,np.newaxis,:]
   146      1216         1804      1.5      0.3          d_sets = data_sets[np.newaxis,:,:]
   147                                                   # import ipdb; ipdb.set_trace()
   148                                                   # for i,data_set in enumerate(data_sets):
   149                                                   #     # temp_array[:,i] = np.sum(self.multi_edge_multisample_loglik_vectorized(
   150                                                   #     #     hidden_states,data_set,gp_out.psi,gp_out.r),axis=1)
   151                                                   #     temp_array[:,i] = np.sum(self.multi_edge_loglik_vectorized(
   152                                                   #         hidden_states,data_set,gp_out.psi,gp_out.r),axis=1)
   153                                                   
   154                                           
   155      1216       579089    476.2     99.1          return np.sum(self.multi_edge_loglik_vectorized(h_states,d_sets,gp_out.psi,gp_out.r),axis=-1)


In [8]:
lstats

In [22]:
[(a,b) for a,b in enumerate(range(5))]

[(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)]

In [16]:
import numpy as np
test = np.array([0,1,2])

In [23]:
np.tile(test,[5,1])
np.tile([0,1,2],[5])

array([0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2])